bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -788,89 +852,340 @@ private long readLastLogId(File f) {
+         */
+        void forceCloseCurrentLogs();
+
+        /*
+         * this method should be called before doing entrymemtable flush, it
+         * would save the state of the entrylogger before entrymemtable flush
+         * and commitEntryMemTableFlush would take appropriate action after
+         * entrymemtable flush.
+         */
+        void prepareEntryMemTableFlush();
+
         /*
-         * In the case of entryLogPerLedgerEnabled we need to flush both
-         * rotatedlogs and currentlogs. This is needed because syncThread
-         * periodically does checkpoint and at this time all the logs should
-         * be flushed.
+         * this method should be called after doing entrymemtable flush,it would
+         * take appropriate action after entrymemtable flush depending on the
+         * current state of the entrylogger and the state of the entrylogger
+         * during prepareEntryMemTableFlush.
          *
-         * TODO: When EntryLogManager is introduced in the subsequent sub-tasks of
-         * this Issue, I will move this logic to individual implamentations of
-         * EntryLogManager and it would be free of this booalen flag based logic.
+         * It is assumed that there would be corresponding
+         * prepareEntryMemTableFlush for every commitEntryMemTableFlush and both
+         * would be called from the same thread.
          *
+         * returns boolean value indicating whether EntryMemTable should do checkpoint
+         * after this commit method.
          */
-        if (entryLogPerLedgerEnabled) {
-            flushCurrentLog();
-        }
+        boolean commitEntryMemTableFlush() throws IOException;
     }
 
-    void flushRotatedLogs() throws IOException {
-        List<BufferedLogChannel> channels = null;
-        long flushedLogId = INVALID_LID;
-        synchronized (this) {
-            channels = logChannelsToFlush;
-            logChannelsToFlush = null;
+    abstract class EntryLogManagerBase implements EntryLogManager {
+        final Set<BufferedLogChannel> rotatedLogChannels;
+
+        EntryLogManagerBase() {
+            rotatedLogChannels = ConcurrentHashMap.newKeySet();
         }
-        if (null == channels) {
-            return;
+
+        /*
+         * This method should be guarded by a lock, so callers of this method
+         * should be in the right scope of the lock.
+         */
+        @Override
+        public long addEntry(Long ledger, ByteBuf entry, boolean rollLog) throws IOException {
+
+            int entrySize = entry.readableBytes() + 4; // Adding 4 bytes to prepend the size
+            BufferedLogChannel logChannel = getCurrentLogForLedger(ledger);
+            boolean reachEntryLogLimit = rollLog ? reachEntryLogLimit(logChannel, entrySize)
+                    : readEntryLogHardLimit(logChannel, entrySize);
+            // Create new log if logSizeLimit reached or current disk is full
+            boolean diskFull = (logChannel == null) ? false
```

Comment:
> it is not much of a change here. There are minor changes here and these changes make code more organized and appropriate.

understand you want to reorganize these. but if possible, can we do this in a subsequent change? it is hard to reason about why it was changed in this way. especially when other people  looking at the git commits in future.

Reply:
> I removed the need of LedgerDirsListener and another flag variable - shouldCreateNewEntryLog. Instead I check if the dir is full by calling "ledgerDirsManager.isDirFull(logChannel.getLogFile().getParentFile())" here. 

I don't think it is a good idea to check disk usage on every add request. This is much expensive than just check `shouldCreateNewEntryLog`. `shouldCreateNewEntryLog` is set when disk checker detect a disk is full, so it is an `asychronous` behavior to avoid any potential performance penalty.

> Also, I don't think it is appropriate to createNewLog in the case of "diskAlmostFull", because though disk is almost full,

createNewLog is *only* set when "current" disk is full at the case of `diskAlmotFull`



The main concern I have with getting rid of shouldCreateNewEntryLog is you change the `async-disk-checking` behavior to a `sync-disk-checking` on *every* add entry. this is unacceptable to me.